internal/impl/git/input.go

```diff
@@ -0,0 +1,550 @@
+// Copyright 2025 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/bmatcuk/doublestar/v4"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	githttp "github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+// Ensure input implements service.Input at compile time.
+var _ service.Input = (*input)(nil)
+
+// input implements a service.Input that reads files from a Git repository.
+// It clones the repository, monitors for changes, and emits file contents as messages.
+type input struct {
+	// cfg contains all config parameters for this input.
+	cfg inputCfg
+	// log is the logger instance for this input.
+	log *service.Logger
+	// filesChan is used to send file details from the scanner to the reader.
+	filesChan chan fileEvent
+	// closeCh signals when the input should stop processing.
+	closeCh chan struct{}
+	// repository is the Git repository instance.
+	repository *git.Repository
+	// lastCommit is the hash of the most recently processed commit.
+	lastCommit plumbing.Hash
+	// tempDir is the temporary directory where the repository is cloned.
+	tempDir string
+}
+
+// fileEvent represents a file change event.
+type fileEvent struct {
+	// path is the absolute path to the file.
+	path string
+	// isDeleted indicates whether the file was deleted.
+	isDeleted bool
+}
+
+// init registers the Git input plugin with the service registry.
+func init() {
+	err := service.RegisterInput(
+		"git", gitInputConfig(),
+		func(parsedCfg *service.ParsedConfig, mgr *service.Resources) (service.Input, error) {
+			conf, err := inputCfgFromParsed(parsedCfg)
+			if err != nil {
+				return nil, err
+			}
+
+			return newInput(conf, mgr)
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// newInput creates a new Git input instance from a parsed configuration.
+func newInput(cfg inputCfg, mgr *service.Resources) (*input, error) {
+	i := &input{
+		cfg:       cfg,
+		filesChan: make(chan fileEvent, 100),
```

Comment:
Why 100? I would suggest just making an unbuffered channel for now unless we see performance issues